rlberry/agents/torch/sac/sac.py

```diff
@@ -0,0 +1,543 @@
+        self.autotune_alpha = autotune_alpha
+        if not self.autotune_alpha:
+            self.alpha = alpha
+
+        # initialize
+        self.reset()
+
+    def reset(self, **kwargs):
+        """
+        Reset the agent.
+        This function resets the agent by initializing the necessary components and parameters for training.
+        """
+
+        # Initialize the rollout buffer
+        self.memory = ReplayBuffer(max_replay_size=self.buffer_capacity, rng=self.rng)
+        self.memory.setup_entry("states", dtype=np.float32)
+        self.memory.setup_entry("next_states", dtype=np.float32)
+        self.memory.setup_entry("actions", dtype=np.float32)
+        self.memory.setup_entry("rewards", dtype=np.float32)
+        self.memory.setup_entry("dones", dtype=np.float32)
+
+        # Intialize the Actor
+        self.cont_policy = self.policy_net_fn(self.env, **self.policy_net_kwargs).to(
+            self.device
+        )
+        self.policy_optimizer = optimizer_factory(
+            self.cont_policy.parameters(), **self.policy_optimizer_kwargs
+        )
+        self.cont_policy.load_state_dict(self.cont_policy.state_dict())
+
+        # Intialize the Q networks and their targets
+        self.q1 = self.q_net_constructor(self.env, **self.q_net_kwargs)
+        self.q2 = self.q_net_constructor(self.env, **self.q_net_kwargs)
+        self.q1_target = self.q_net_constructor(self.env, **self.q_net_kwargs)
+        self.q2_target = self.q_net_constructor(self.env, **self.q_net_kwargs)
+        self.q1.to(self.device)
+        self.q2.to(self.device)
+        self.q1_target.to(self.device)
+        self.q2_target.to(self.device)
+        self.q1_optimizer = optimizer_factory(
+            self.q1.parameters(), **self.q_optimizer_kwargs
+        )
+        self.q2_optimizer = optimizer_factory(
+            self.q2.parameters(), **self.q_optimizer_kwargs
+        )
+        self.q1_target_optimizer = optimizer_factory(
+            self.q1.parameters(), **self.q_optimizer_kwargs
+        )
+        self.q2_target_optimizer = optimizer_factory(
+            self.q2.parameters(), **self.q_optimizer_kwargs
+        )
+        # Define the loss
+        self.MseLoss = nn.MSELoss()
+
+        # Automatic entropy tuning
+        if self.autotune_alpha:
+            self.target_entropy = -torch.prod(
+                torch.Tensor(self.env.action_space.shape).to(self.device)
+            ).item()
+            self.log_alpha = torch.zeros(1, requires_grad=True, device=self.device)
+            self.alpha = self.log_alpha.exp().item()
+            self.a_optimizer = optim.Adam([self.log_alpha], lr=self.q_learning_rate)
+
+        # initialize episode, steps and time counters
+        self.total_episodes = 0
+        self.total_timesteps = 0
+        self.time = time.time()
+
+    def policy(self, state):
+        assert self.cont_policy is not None
+        state = np.array([state])
+        state = torch.FloatTensor(state).to(self.device)
```

Comment:
Just a nit, but there is some built-in Pytorch method to convert a np array to a torch tensor (`torch.as_tensor`, `torch.from_numpy`). I think `torch.as_tensor` allow you to specify the device directly.

Reply:
Thanks for the suggestions Waris,  I tried here to be consistent with the other implementations of Rlberry, `FloatTensor` was used in some of them, but so was `torch.from_numpy`. That's why I kept it. Is there any difference in terms of performance otherwise ?